C#/Botkit/Microsoft.BotKit/BotWorker.cs

```diff
@@ -123,25 +144,129 @@ public async Task<DialogTurnResult> CancelAllDialogs()
+        public async Task ReplaceDialogAsync(string id, Dictionary<string, object> options)
         {
+            if (config.DialogContext == null)
+            {
+                throw new Exception("Call to beginDialog on a bot that did not receive a dialogContext during spawn");
+            }
+
+            var opt = new Dictionary<string, object>();
+
+            // TO-DO: Review this implementation of 'Options' as Dictionary
+            foreach (KeyValuePair<string, object> entry in options)
+            {
+                opt.Add(entry.Key, entry.Value);
+            }
 
+            opt.Add("user", config.TurnContext.Activity.From.Id);
+            opt.Add("channel", config.TurnContext.Activity.Conversation.Id);
+
+            await config.DialogContext.ReplaceDialogAsync(id + ":botkit-wrapper", opt);
+
+            // make sure we save the state change caused by the dialog.
+            // this may also get saved again at end of turn
+            await Controller.SaveState(this);
         }
 
         /// <summary>
         /// Alter the context in which a bot instance will send messages.
         /// Use this method to create or adjust a bot instance so that it can send messages to a predefined user/channel combination.
         /// </summary>
-        /// <param name="reference">A ConversationReference, most likely captured from an incoming message and stored for use in proactive messaging scenarios.</param>
-        public async Task<BotWorker> ChangeContext(ConversationReference reference)
+        /// <param name="conversationReference">A ConversationReference, most likely captured from an incoming message and stored for use in proactive messaging scenarios.</param>
+        public async Task<BotWorker> ChangeContextAsync(ConversationReference conversationReference)
         {
-            return new BotWorker(new Botkit(new BotkitConfiguration()), new object());
-        }
+            // change context of outbound activities to use this new address
+            config.ConversationReference = conversationReference;
 
+            // Create an activity using this reference
+            var activity = new Activity();
+            activity = activity.ApplyConversationReference(conversationReference, true);
 
-        public async void StartConversationWithUser(object reference)
+            // create a turn context
+            var turnContext = new TurnContext(Controller.Adapter, activity);
+
+            // create a new dialogContext so beginDialog works.
+            var dialogContext = await Controller.DialogSet.CreateContextAsync(turnContext);
+
+            config.TurnContext = turnContext;
+            config.DialogContext = dialogContext;
+            config.Activity = activity;
+
+            return this;
+        }
+
+        public async Task StartConversationWithUserAsync(ConversationReference convReference)
         {
+            // this code is mostly copied from BotFrameworkAdapter.createConversation
```

Comment:
you sure?
https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs#L743

Reply:
Removed 👍